src/tree_data.c

```diff
@@ -280,117 +282,171 @@ lyd_meta2str(const struct lyd_meta *meta, int *dynamic)
-#if 0
-    const char *yang_data_name = NULL;
-#endif
 
-    LY_CHECK_ARG_RET(ctx, ctx, NULL);
+    if (!format && in->type == LY_IN_FILEPATH) {
+        /* unknown format - try to detect it from filename's suffix */
+        const char *path = in->method.fpath.filepath;
+        size_t len = strlen(path);
 
-    if ((options & LYD_OPT_PARSE_ONLY) && (options & LYD_VALOPT_MASK)) {
-        LOGERR(ctx, LY_EINVAL, "Passing validation flags with LYD_OPT_PARSE_ONLY is not allowed.");
-        return NULL;
-    }
+        /* ignore trailing whitespaces */
+        for (; len > 0 && isspace(path[len - 1]); len--);
 
+        if (len >= 5 && !strncmp(&path[len - 4], ".xml", 4)) {
+            format = LYD_XML;
 #if 0
-    if (options & LYD_OPT_RPCREPLY) {
-        /* first item in trees is mandatory - the RPC/action request */
-        LY_CHECK_ARG_RET(ctx, trees, LY_ARRAY_SIZE(trees) >= 1, NULL);
-        if (!trees[0] || trees[0]->parent || !(trees[0]->schema->nodetype & (LYS_ACTION | LYS_LIST | LYS_CONTAINER))) {
-            LOGERR(ctx, LY_EINVAL, "Data parser invalid argument trees - the first item in the array must be the RPC/action request when parsing %s.",
-                   lyd_parse_options_type2str(options));
-            return NULL;
-        }
+        } else if (len >= 6 && !strncmp(&path[len - 5], ".json", 5)) {
+            format = LYD_JSON;
+#endif
+        } else if (len >= 5 && !strncmp(&path[len - 4], ".lyb", 4)) {
+            format = LYD_LYB;
+        } /* else still unknown */
     }
 
-    if (options & LYD_OPT_DATA_TEMPLATE) {
-        yang_data_name = va_arg(ap, const char *);
-    }
-#endif
+    return format;
+}
 
-    if (!format) {
-        /* TODO try to detect format from the content */
-    }
+API LY_ERR
+lyd_parse_data(const struct ly_ctx *ctx, struct ly_in *in, LYD_FORMAT format, int parse_options, int validate_options, struct lyd_node **tree)
+{
+    LY_CHECK_ARG_RET(ctx, ctx, in, tree, LY_EINVAL);
+    LY_CHECK_ARG_RET(ctx, !(parse_options & ~LYD_PARSE_OPTS_MASK), LY_EINVAL);
+    LY_CHECK_ARG_RET(ctx, !(validate_options & ~LYD_VALIDATE_OPTS_MASK), LY_EINVAL);
+
+    format = lyd_parse_get_format(in, format);
+    LY_CHECK_ARG_RET(ctx, format, LY_EINVAL);
+
+    LY_CHECK_ARG_RET(ctx, format, LY_EINVAL);
 
     switch (format) {
     case LYD_XML:
-        lyd_parse_xml_data(ctx, data, options, &result);
-        break;
+        return lyd_parse_xml_data(ctx, in->current, parse_options, validate_options, tree);
 #if 0
     case LYD_JSON:
-        lyd_parse_json(ctx, data, options, trees, &result);
-        break;
+        return lyd_parse_json_data(ctx, in->current, parse_options, validate_options, tree);
 #endif
     case LYD_LYB:
-        lyd_parse_lyb_data(ctx, data, options, &result, NULL);
-        break;
+        return lyd_parse_lyb_data(ctx, in->current, parse_options, validate_options, tree, NULL);
     case LYD_SCHEMA:
-        LOGINT(ctx);
-        break;
+        LOGINT_RET(ctx);
```

Comment:
There should not be a return yet, but final top-level validation should be performed here for all the formats. Meaning move the `if` block at `parser_xml.c:698` here and remove `TODO` in `parser_lyb.c:1060`.

Reply:
The joint top-level validation for all the data formats parsers will be done later together with the JSON format data parser.